File: Python/findMaximizedCapital.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: findMaximizedCapital.py
@Function: IPO
@Link: https://leetcode-cn.com/problems/ipo/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-25
"""


from typing import List
from heapq import nlargest, heappop, heappush


class Solution:
    def findMaximizedCapital(self, k: int, W: int, Profits: List[int], Capital: List[int]) -> int:
        # to speed up: if all projects are available
        if W >= max(Capital):
            return W + sum(nlargest(k, Profits))

        n = len(Profits)
        projects = [(Capital[i], Profits[i]) for i in range(n)]
        # sort the projects :
        # the most available (= the smallest capital) is the last one
        projects.sort(key=lambda x: -x[0])

        available = []
        while k > 0:
            # update available projects
            while projects and projects[-1][0] <= W:
                heappush(available, -projects.pop()[1])
            # if there are available projects,
            # pick the most profitable one
            if available:
                W -= heappop(available)
            # not enough capital to start any project
            else:
                break
            k -= 1
        return W


# -------------------------
if __name__ == "__main__":
    k = 2
    W = 0
    Profits = [1, 2, 3]
    Capital = [0, 1, 1]

    solution = Solution()
    find_max_min_capial = solution.findMaximizedCapital(k, W, Profits, Capital)
    print(f"The solution of this problem is {find_max_min_capial}")

File: C++/findMaximizedCapital.cpp
/**
 * @File    : findMaximizedCapital.cpp
 * @Brief   : IPO
 * @Link    : https://leetcode-cn.com/problems/ipo/
 * @Author  : Wei Li
 * @Date    : 2021-08-25
*/

#include <iostream>
#include <vector>
#include <utility>
#include <queue>
#include <algorithm>

class Solution
{
public:
    int findMaximizedCapital(int k, int w, std::vector<int> &profits, std::vector<int> &capital)
    {
        std::vector<std::pair<int, int>> capital2profits;
        for (int i = 0; i < profits.size(); i++)
        {
            capital2profits.push_back({capital[i], profits[i]});
        }
        auto cmp = [&](std::pair<int, int> &a, std::pair<int, int> &b)
        {
            return a.first < b.first;
        };
        std::sort(capital2profits.begin(), capital2profits.end(), cmp);
        int haveMoney = w;
        std::priority_queue<int> maxProfit;
        int i = 0;
        while (k--)
        {
            for (; i < capital2profits.size(); i++)
            {
                if (capital2profits[i].first <= haveMoney)
                {
                    maxProfit.push(capital2profits[i].second);
                }
                else
                    break;
            }
            if (maxProfit.empty())
                break;
            haveMoney += maxProfit.top();
            maxProfit.pop();
        }
        return haveMoney;
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    int k = 2, w = 0;
    std::vector<int> profits{1, 2, 3};
    std::vector<int> captial{0, 1, 1};

    Solution solution;
    auto find_max_min_capial = solution.findMaximizedCapital(k, w, profits, captial);
    std::cout << "The soluton of this problem is " << find_max_min_capial << std::endl;

    return 0;
}
